### DISCUSSION TITLE
[MacOS] rocks.nvim does not work after successful installation with luajit

### DISCUSSION THREAD

**Discussion by hpb0412** [lumen-oss/rocks.nvim#385] [Q&A]
I installed `rocks.nvim` using the recommended option at README.md

The installation is successful, no error.

I followed the instructions and put the generated code into `~/.config/nvim/init.lua`

However, I cannot install any package. Ex:
```shell
:Rocks install kanagawa.nvim
```
return these errors in `:Rocks log`
```shell
[START][2024-06-15 16:04:23] rocks.nvim logging initiated
ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:65 | env: lua: No such file or directory

ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:65 | env: lua: No such file or directory

ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-15 16:04:25 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:65 | env: lua: No such file or directory

ERROR | 2024-06-15 16:07:47 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-15 16:07:47 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:65 | env: lua: No such file or directory

ERROR | 2024-06-15 16:08:09 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:64 | luarocks CLI FAILED
ERROR | 2024-06-15 16:08:09 | ....local/share/nvim/rocks/share/lua/5.1/rocks/luarocks.lua:65 | env: lua: No such file or directory

ERROR | 2024-06-15 16:08:09 | ...re/nvim/rocks/share/lua/5.1/rocks/operations/helpers.lua:69 | Failed to install kanagawa.nvim
```
and every time I open `nvim`, I got this error
<img width="1512" alt="Screen Shot 2024-06-15 at 4 05 10 PM" src="https://github.com/nvim-neorocks/rocks.nvim/assets/7277539/ad71b299-842c-441c-8998-db306b8f7737">

Here is my environment

```shell
nvim -v
NVIM v0.10.0
Build type: Release
LuaJIT 2.1.171665647

luajit -v
LuaJIT 2.1.1716656478 -- Copyright (C) 2005-2023 Mike Pall. https://luajit.org/

git -v
git version 2.39.0

curl --version
curl 8.1.2 (x86_64-apple-darwin21.0) libcurl/8.1.2 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.11 nghttp2/1.45.1
Release-Date: 2023-05-30
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
```

```lua
:lua vim.fn.executable("luarocks") -- return 1
```

```shell
# :checkhealth

rocks: require("rocks.health").check()

Checking external dependencies ~
- OK luarocks: found 
- WARNING         lua: not found.
          Install [Lua](https://www.lua.org/) for extended capabilities.
          luarocks requires a Lua installation.
          

Checking rocks.nvim config ~
- OK No errors found in config.

Checking rocks.toml ~
- OK No errors found in rocks.toml.
```

**Comment by mrcjkb** [comment]
Hey :wave: 

As the health check and error messages indicate, luarocks expects a `lua` executable.
You have `luajit` installed, and from the looks of it, it doesn't have a `lua` alias or symlink.

**Reply by hpb0412** [reply]
Thank you for quick reply!

I thought I only need `luajit` since the readme mentioned:

```md
A lua 5.1 or luajit installation (for luarocks).
```

And `:lua vim.fn.executable("luarocks")` returns `1` also supports that thought.

Anyway, let me install `lua 5.1`  and feedback later.

**Reply by mrcjkb** [reply]
Yep, luajit should work, but it needs to have a `lua` alias. 

Unfortunately, `vim.fn.executable("luarocks")` returning `1`  only indicates that an executable exists, but not that it works.

> let me install `lua 5.1` and feedback later.

This might not be possible on MacOS if you're using Homebrew.
But creating an `alias lua luajit` might be enough.

@craigmac you [added the readme note about using `luajit` on MacOS](https://github.com/nvim-neorocks/rocks.nvim/pull/363).
Do you know if there's an extra step needed to create an alias?

**Reply by hpb0412** [reply]
@mrcjkb I tried both. Let me provide info here:

- I installed `lua@5.4` using `brew install lua` => `rocks.nvim` works (at least basic functions, install, remove, sync)
- (Remove `lua` installed above) I set alias with `alias lua=luajit` => `rocks.nvim` does not work (with different msg):
```shell
which lua
lua: aliased to luajit
```
<img width="1512" alt="image" src="https://github.com/nvim-neorocks/rocks.nvim/assets/7277539/87f119f5-36a1-4ea8-a7e9-832576ca8250">

**Reply by hpb0412** [reply]
Also, small update:

From current status (remove lua, set alias => not work)

I remove the alias, reinstall lua but again I still get errors, no matter if I
- restart the terminal
- restart the machine

The only way to fix it is to remove these directories and reinstall the `rocks.nvim`
- `~/.local/share/nvim/rocks`
- `~/.local/share/nvim/site/pack/luarocks`

**Reply by mrcjkb** [reply]
It looks like you were running into #377, which should be fixed with the latest rocks.nvim release.

**Comment by hpb0412** [comment]
Another update:

I hope that `lua@5.4` will work but it's failed when I try to install `tree-sitter-{lang}`.

I thought it was the end as you know `brew` has not allowed installing `lua@5.1` from @craigmac's PR.

Fortunately, I have `asdf` to install `lua@5.1` instead of `brew`. Now, `rocks.nvim` seems to work without glitches.

At this point, I guess if the `alias` solution was not working for me, it might not be because of missing steps, but because of version mismatching. Since `lua` must be at `5.1`, then `luajit` should be at a specific version, right?

Anyway, I hope the above information will help others if they fall into this situation.

**Comment by craigmac** [comment]
Since, lua@51 has been disabled, you have two options, you can try the brew extract method which means you extract that disabled formula to your own 'tap' and install and maintain it yourself, it's a bit involved though. I opted instead to just download lua 5.1 and install it from source doing:

```
cd ~/Downloads
curl -O https://www.lua.org/ftp/lua-5.1.5.tar.gz
tar xvf lua-5.1.5.tar.gz
cd lua-5.1.5
make macosx
sudo make install
```

Will install into `/usr/local/bin/lua` where it will be found properly by rocks.nvim

**Comment by andrewcrook** [comment]
Don't forget builds will use the lua version available this sometimes hard codes packages etc  to a specific version of lua.
So if you change the lua version your will have to reinstall in order to rebuild using the different lua version. 

On a side note, I don't think lua5.1 and Luajit  should be side by side (I have had several issues where things pickup the wrong version ) really you want to use one at a time in a shell, one of the other may be use some sort of shim script?

**Reply by mrcjkb** [reply]
Thanks for the input.

> I have had several issues where things pickup the wrong version

Have you had these issues with neovim and/or rocks.nvim? Usually, Neovim comes bundled with a luajit interpreter, which should be able to run native Lua built against either luajit or Lua 5.1 headers.

To make sure luarocks doesn't pick up Lua > 5.1 (e.g. 5.4), rocks.nvim passes `--lua-version 5.1` to the `luarocks` CLI.

**Reply by andrewcrook** [reply]
> Have you had these issues with neovim and/or rocks.nvim
Only when using mason it was ignoring asdf and using luajit from homebrew
I removed the plugin from asdf and it started using lua 5.1 correctly. I am not sure what was causing it.
I use multiple versions with asdf but the lua and luajit plugins are broken and luajit one is out of date. I plan to rewrite them as one.

**Reply by mrcjkb** [reply]
Oh yeah, mason can be finicky.
